docs/proposals/resource-metrics-api.md

```diff
@@ -0,0 +1,177 @@
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Resource Metrics API
+
+*This proposal is based on and supersedes [compute-resource-metrics-api.md](compute-resource-metrics-api.md).*
+
+This document describes API part of MVP version of Resource Metrics API effort in Kubernetes.
+Once the agreement will be made the document will be extended to also cover implementation details.
+The shape of the effort may be also a subject of changes once we will have more well-defined use cases.
+
+## Goal
+
+The goal for the effort is to provide resource usage metrics for pods and nodes through the API server.
+This will be a stable, versioned API which core Kubernetes components can rely on.
+In the first version only the well-defined use cases will be handled,
+altough the API should be easily extensible for potential future use cases.
+
+## Main use cases
+
+This section describes well-defined use cases which should be handled in the first version.
+Use cases which are not listed below are out of the scope of MVP version of Resource Metrics API.
+
+#### Horizontal Pod Autoscaler
+
+HPA uses the latest value of cpu usage as an average aggregated across 1 minute
+(the window may change in the future). The data for a given set of pods
+(defined either by pod list or label selector) should be accesible in one request
+due to performance issues.
+
+#### Scheduler
+
+Scheduler in order to schedule best-effort pods requires node level resource usage metrics
+as an average aggreated across 1 minute (the window may change in the future).
+The metrics should be available for all resources supported in the scheduler.
+Currently the scheduler does not need this information, because it schedules best-effort pods
+without considering node usage. But having the metrics available in the API server is a blocker
+for adding the ability to take node usage into account when scheduling best-effort pods.
+
+## Other considered use cases
+
+This section describes the other considered use cases and explains why they are out
+of the scope of the MVP version.
+
+#### Custom metrics in HPA
+
+HPA requires the latest value of application level metrics.
+
+The design of the pipeline for collecting application level metrics should
+be revisited and it's not clear whether application level metrics should be
+available in API server so the use case initially won't be supported.
+
+#### Ubernetes
+
+Ubernetes might want to consider cluster-level usage (in addition to cluster-level request)
+of running pods when choosing where to schedule new pods. Although Ubernetes is still in design,
+we expect the metrics API described here to be sufficient. Cluster-level usage can be
+obtained by summing over usage of all nodes in the cluster.
```

Comment:
Nitpicking: for cluster-wide rank statistics such as 95th and 99th percentile, and people will want them, you can't compute those from pre-aggregated data. Think of the case where you look at the max over a window of one hour and a large RC/RS moves from one set of machines to another: it will be counted twice. (You also need to decide whether it's cluster-wide, but computed at the container, pod or node level... most likely people will want the last two.) One way to approximate this would be computing something like Q-digests or t-digests from data streamed from PodMetrics (not node data). Unless you only want to support instantaneous usage, the last sentence is misleading.